File: backend/pkg/core/game_stats.go
package core

import "time"

// GameStats is the full post-game statistics payload for a completed game.
//
// Terminology note: what players call a "post" is a `comment` row in the
// messages table. The `post` message_type is the GM-authored topic container a
// phase's discussion hangs off (typically one post per phase with hundreds of
// comments beneath it). Every writing statistic here is therefore computed over
// comments; GM posts are deliberately excluded.
type GameStats struct {
	Comments        CommentStats        `json:"comments"`
	PrivateMessages PrivateMessageStats `json:"private_messages"`
	Characters      []CharacterWriting  `json:"characters"`
	// LongestConversation is nil when the game has no private conversations
	// carrying any messages, which is a normal state and not an error.
	LongestConversation *LongestConversation `json:"longest_conversation,omitempty"`
	// LongestComment is nil when the game has no comments.
	LongestComment *LongestComment `json:"longest_comment,omitempty"`
}

// LongestComment identifies the single longest comment so the UI can deep-link
// to it. Separate from CommentStats.LongestCommentLength, which is the plain
// aggregate; this carries the identity needed to build the link.
type LongestComment struct {
	MessageID     int32  `json:"message_id"`
	ContentLength int64  `json:"content_length"`
	CharacterName string `json:"character_name,omitempty"`
	// PhaseID is the phase the comment belongs to, used to resolve the deep
	// link in the History view. Zero when the comment has no phase.
	PhaseID int32 `json:"phase_id,omitempty"`
}

// CommentStats holds game-wide common room aggregates.
type CommentStats struct {
	TotalComments        int64 `json:"total_comments"`
	AvgCommentLength     int64 `json:"avg_comment_length"`
	LongestCommentLength int64 `json:"longest_comment_length"`
	// MaxThreadDepth and AvgThreadDepth describe reply-tree shape. Depth is
	// stored 0-based on messages.thread_depth.
	MaxThreadDepth int64   `json:"max_thread_depth"`
	AvgThreadDepth float64 `json:"avg_thread_depth"`
}

// PrivateMessageStats holds game-wide private messaging aggregates.
type PrivateMessageStats struct {
	TotalPrivateMessages        int64 `json:"total_private_messages"`
	AvgPrivateMessageLength     int64 `json:"avg_private_message_length"`
	LongestPrivateMessageLength int64 `json:"longest_private_message_length"`
	// ActiveConversations counts conversations that carry at least one
	// non-deleted message, not conversations that were merely created.
	ActiveConversations int64 `json:"active_conversations"`
}

// LongestConversation describes the private conversation with the most
// messages. "Longest" is defined by message count rather than total characters
// written or wall-clock duration.
type LongestConversation struct {
	ConversationID   int32      `json:"conversation_id"`
	Title            string     `json:"title,omitempty"`
	ConversationType string     `json:"conversation_type"`
	MessageCount     int64      `json:"message_count"`
	ParticipantNames []string   `json:"participant_names"`
	FirstMessageAt   *time.Time `json:"first_message_at,omitempty"`
	LastMessageAt    *time.Time `json:"last_message_at,omitempty"`
}

// CharacterWriting holds one character's writing stats within a game.
// Characters with no activity are still included, with zero counts.
type CharacterWriting struct {
	CharacterID             int32  `json:"character_id"`
	CharacterName           string `json:"character_name"`
	IsActive                bool   `json:"is_active"`
	CommentCount            int64  `json:"comment_count"`
	AvgCommentLength        int64  `json:"avg_comment_length"`
	PrivateMessageCount     int64  `json:"private_message_count"`
	AvgPrivateMessageLength int64  `json:"avg_private_message_length"`
}

File: frontend/src/types/gameStats.ts
/**
 * Post-game statistics for completed games.
 *
 * Terminology: a "comment" here is what players call a post — an entry in a
 * common room discussion. The backend's `post` message type is the GM-authored
 * topic each phase's discussion hangs off, and is excluded from every statistic
 * below. The UI says "comments" throughout to keep the two unambiguous.
 */

interface CommentStats {
  total_comments: number;
  avg_comment_length: number;
  longest_comment_length: number;
  /** Reply nesting depth, 0-based on the topic post. */
  max_thread_depth: number;
  avg_thread_depth: number;
}

interface PrivateMessageStats {
  total_private_messages: number;
  avg_private_message_length: number;
  longest_private_message_length: number;
  /** Conversations carrying at least one message, not merely created. */
  active_conversations: number;
}

/** The private conversation with the most messages. */
interface LongestConversation {
  conversation_id: number;
  title?: string;
  conversation_type: string;
  message_count: number;
  participant_names: string[];
  first_message_at?: string;
  last_message_at?: string;
}

/** One character's writing stats. Characters with no activity appear with zeros. */
export interface CharacterWriting {
  character_id: number;
  character_name: string;
  is_active: boolean;
  comment_count: number;
  avg_comment_length: number;
  private_message_count: number;
  avg_private_message_length: number;
}

/** The single longest comment, carrying enough identity to deep-link to it. */
interface LongestComment {
  message_id: number;
  content_length: number;
  character_name?: string;
  phase_id?: number;
}

export interface GameStats {
  comments: CommentStats;
  private_messages: PrivateMessageStats;
  characters: CharacterWriting[];
  /** Absent when the game had no private conversations carrying messages. */
  longest_conversation?: LongestConversation;
  /** Absent when the game had no comments. */
  longest_comment?: LongestComment;
}
